File: Dddml.Wms.JavaCommon/src/main/java/org/dddml/wms/specialization/ApplicationContext.java
package org.dddml.wms.specialization;


/**
 * Created by Yang on 2016/7/20.
 */
public class ApplicationContext {

    public static volatile ApplicationContext current;

    public static void setCurrent(ApplicationContext context) {
        current = context;
    }

    public Object get(String name) {
        throw new UnsupportedOperationException();
    }


    public TypeConverter getTypeConverter() {
        return new DefaultTypeConverter();
    }

    public static class DefaultTypeConverter implements TypeConverter {

        private org.dddml.support.criterion.DefaultTypeConverter innerTypeConverter = new org.dddml.support.criterion.DefaultTypeConverter();

        @Override
        public Object convertFromString(Class type, String text) {
            return innerTypeConverter.convertFromString(type, text);
        }

        @Override
        public String convertToString(Object value) {
            return innerTypeConverter.convertToString(value);
        }

        @Override
        public String convertToString(Class type, Object value) {
            return innerTypeConverter.convertToString(value);
        }

        @Override
        public String[] convertToStringArray(Object[] values) {
            return innerTypeConverter.convertToStringArray(values);
        }
    }
}

File: Dddml.Wms.Specialization/Specialization/ApplicationContext.cs
using System;


namespace Dddml.Wms.Specialization
{
    public abstract class ApplicationContext : IApplicationContext
    {

        private static volatile ApplicationContext _current;

        public static ApplicationContext Current
        {
            get
            {
                return _current;
            }
            set
            {
                _current = value;
            }
        }

        private ITypeConverter _typeConverter = new DefaultTypeConverter();

        public virtual ITypeConverter TypeConverter
        {
            get { return _typeConverter; }
            set { _typeConverter = value; }
        }

        public abstract object this[string name]
        {
            get;
        }

        public virtual IClobConverter ClobConverter
        {
            get
            {
                return (IClobConverter)this["ClobConverter"];
            }
        }

        public virtual ITimestampService TimestampService
        {
            get { return new DefaultTimestampService(); }
        }

        public class DefaultTimestampService : ITimestampService
        {
            public T Now<T>()
            {
                if (typeof(T).Equals(typeof(DateTime)))
                {
                    return (T)(object)DateTime.Now;
                }
                else if (typeof(T).Equals(typeof(long)))
                {
                    return (T)(object)DateTime.Now.Ticks;
                }
                else if (typeof(T).Equals(typeof(DateTime?)))
                {
                    return (T)(object)new DateTime?(DateTime.Now);
                }
                else if (typeof(T).Equals(typeof(long?)))
                {
                    return (T)(object)new long?(DateTime.Now.Ticks);
                }

                throw new ArgumentException("Unknown type: " + typeof(T));
            }

        }

    }
}

